source/docs/flow_control.md

```diff
@@ -0,0 +1,183 @@
+### Overview
+
+Flow control in Envoy is done by having limits on each buffer, and watermark callbacks.  When a
+buffer contains more data than the configured limit, the high watermark callback will fire, kicking
+off a chain of events which eventually informs the data source to stop sending data.  This back-off
+may be immediate (stop reading from a socket) or gradual (stop HTTP/2 window updates) so all
+buffer limits in Envoy are considered soft limits.  When the buffer eventually drains (generally to
+half of of the high watermark to avoid thrashing back and forth) the low watermark callback will
+fire, informing the sender it can resume sending data.
+
+### TCP implementation details
+
+Flow control for TCP and TCP-with-TLS-termination are handled by coordination
+between the `Network::ConnectionImpl` write buffer, and the `Network::TcpProxy`
+filter.
+
+The downstream flow control goes as follows.
+
+ * The downstream `Network::ConnectionImpl::write_buffer_` buffers too much
+   data.  It calls
+   `Network::ConnectionCallbacks::onAboveWriteBufferHighWatermark()`.
+ * The `Network::TcpProxy::DownstreamCallbacks` receives
+   `onAboveWriteBufferHighWatermark()` and calls `readDisable(true)` on the upstream
+   connection.
+ * When the downstream buffer is drained, it calls
+   `Network::ConnectionCallbacks::onBelowWriteBufferLowWatermark()`
+ * The `Network::TcpProxy::DownstreamCallbacks` receives
+   `onBelowWriteBufferLowWatermark()` and calls `readDisable(false)` on the upstream
+   connection.
+
+Flow control for the upstream path is much the same.
+
+ * The upstream `Network::ConnectionImpl::write_buffer_` buffers too much
+   data.  It calls
+   `Network::ConnectionCallbacks::onAboveWriteBufferHighWatermark()`.
+ * The Network::TcpProxy::UpstreamCallbacks receives
+   `onAboveWriteBufferHighWatermark()` and calls `readDisable(true)` on the downstream
+   connection.
+ * When the upstream buffer is drained, it calls
+   `Network::ConnectionCallbacks::onBelowWriteBufferLowWatermark()`
+ * The `Network::TcpProxy::UpstreamCallbacks` receives
+   `onBelowWriteBufferLowWatermark()` and calls `readDisable(false)` on the downstream
+   connection.
+
+### HTTP2 implementation details
+
+Because the various buffers in the HTTP/2 stack are fairly complicated, each path from a buffer
+going over the watermark limit to disabling data from the data source is documented separately.
+
+![HTTP2 data flow diagram](h2_buffers.png)
```

Comment:
Is there a source vector drawing that was used to generate this image?

I think that the color or some of the buffers in the diagram may need to be updated.  I think that based on how L4 buffering works today the ConnectionImpl buffer is bounded by the configured connection watermark and should be green.  Some other changes are also in the pipeline.

Reply:
This comes from https://docs.google.com/document/d/1EB3ybx3yTndp158c4AdQ4nutksZA9lL-BQvixhPnb_4/edit#bookmark=id.umrpbw846kb. I'll share the doc with you and you can copy from there if you want.